### PR TITLE
Initialize hotRootsBySlotCache on startup

### PR DESCRIPTION
**Pull request by cemozerr**
## PR Description
Initialize hotRootsBySlotCache on startup so that we don't miss blocks to prune.

### CODE REVIEW

File: storage/src/test/java/tech/pegasys/teku/storage/server/rocksdb/V3RocksDbDatabaseTest.java [hunk 1]
```diff
@@ -51,6 +55,18 @@ public void shouldHandleRestartWithUnrecoverableForkBlocks_prune(@TempDir final
     testShouldHandleRestartWithUnrecoverableForkBlocks(tempDir, StateStorageMode.PRUNE);
   }
 
+  @Test
+  public void ShouldPruneHotBlocksOlderThanFinalizedSlotAfterRestart__archive(
```
Comment (mbaxter) [inline storage/src/test/java/tech/pegasys/teku/storage/server/rocksdb/V3RocksDbDatabaseTest.java]:
```suggestion
  public void shouldPruneHotBlocksOlderThanFinalizedSlotAfterRestart__archive(
```

File: storage/src/test/java/tech/pegasys/teku/storage/server/rocksdb/V3RocksDbDatabaseTest.java [hunk 2]
```diff
@@ -51,6 +55,18 @@ public void shouldHandleRestartWithUnrecoverableForkBlocks_prune(@TempDir final
...
+      @TempDir final Path tempDir) throws Exception {
+    testShouldPruneHotBlocksOlderThanFinalizedSlotAfterRestart(tempDir, StateStorageMode.ARCHIVE);
+  }
+
+  @Test
+  public void ShouldPruneHotBlocksOlderThanFinalizedSlotAfterRestart__prune(
```
Comment (mbaxter) [inline storage/src/test/java/tech/pegasys/teku/storage/server/rocksdb/V3RocksDbDatabaseTest.java]:
```suggestion
  public void shouldPruneHotBlocksOlderThanFinalizedSlotAfterRestart__prune(
```

File: storage/src/main/java/tech/pegasys/teku/storage/server/rocksdb/dataaccess/V3RocksDbDao.java [hunk 3]
```diff
@@ -173,11 +174,33 @@ private void initialize() {
     }
 
     LOG.info("Initializing hot states from hot blocks");
-    initializeHotStates(finalizedRoot.get(), finalizedState.get(), hotBlocksByRoot);
+    final Map<Bytes32, SignedBeaconBlock> prunedHotBlocks =
+        initializeHotStatesAndBlocks(finalizedRoot.get(), finalizedState.get(), hotBlocksByRoot);
     LOG.info("Finished initializing hot states from hot blocks");
+
+    LOG.info("Adding hot roots to cache ordered by slots");
```
Comment (mbaxter) [inline storage/src/main/java/tech/pegasys/teku/storage/server/rocksdb/dataaccess/V3RocksDbDao.java]: This doesn't take a lot of time, so I don't think it's really worth logging the start / end of the cache initialization.

File: storage/src/test/java/tech/pegasys/teku/storage/server/rocksdb/V3RocksDbDatabaseTest.java [hunk 4]
```diff
@@ -98,4 +114,53 @@ private void testShouldHandleRestartWithUnrecoverableForkBlocks(
     assertStatesUnavailable(unavailableBlockRoots);
     assertBlocksUnavailable(unavailableBlockRoots);
   }
+
+  private void testShouldPruneHotBlocksOlderThanFinalizedSlotAfterRestart(
+      @TempDir final Path tempDir, final StateStorageMode storageMode) throws Exception {
+    // Setup chains
+    final ChainBuilder chain = ChainBuilder.create(VALIDATOR_KEYS);
+    final SignedBlockAndState genesis = chain.generateGenesis();
+    // Primary chain's next block is at 7
```
Comment (mbaxter) [inline storage/src/test/java/tech/pegasys/teku/storage/server/rocksdb/V3RocksDbDatabaseTest.java]: Stale comment
```suggestion
```

File: storage/src/test/java/tech/pegasys/teku/storage/server/rocksdb/V3RocksDbDatabaseTest.java [hunk 5]
```diff
@@ -98,4 +114,53 @@ private void testShouldHandleRestartWithUnrecoverableForkBlocks(
...
+    chain.generateBlocksUpToSlot(10);
+
+    // Setup database
+    database = setupDatabase(tempDir.toFile(), storageMode);
+    store = Store.getForkChoiceStore(genesis.getState());
+    database.storeGenesis(store);
+
+    // Finalize at slot 15, so all the blocks before 15 would need to be pruned.
```
Comment (mbaxter) [inline storage/src/test/java/tech/pegasys/teku/storage/server/rocksdb/V3RocksDbDatabaseTest.java]: Stale comment
```suggestion
```

File: storage/src/test/java/tech/pegasys/teku/storage/server/rocksdb/V3RocksDbDatabaseTest.java [hunk 6]
```diff
@@ -98,4 +114,53 @@ private void testShouldHandleRestartWithUnrecoverableForkBlocks(
...
+
+    final Checkpoint finalizedCheckpoint = getCheckpointForBlock(chain.getBlockAtSlot(7));
+    finalizeCheckpoint(finalizedCheckpoint);
+
+    // We should be able to access hot blocks and state
+    final List<SignedBlockAndState> expectedHotBlocksAndStates =
+        chain.streamBlocksAndStates(7, 10).collect(toList());
+    assertHotBlocksAndStatesInclude(expectedHotBlocksAndStates);
+
+    // Old states should be unavailable
```
Comment (mbaxter) [inline storage/src/test/java/tech/pegasys/teku/storage/server/rocksdb/V3RocksDbDatabaseTest.java]: Stale comment
```suggestion
```

**Review by mbaxter — APPROVED** [review]
LGTM - just a few clean-up-related comments

File: storage/src/test/java/tech/pegasys/teku/storage/server/rocksdb/V3RocksDbDatabaseTest.java (hunk 1 above)
Comment (cemozerr) [reply storage/src/test/java/tech/pegasys/teku/storage/server/rocksdb/V3RocksDbDatabaseTest.java]: done.

File: storage/src/test/java/tech/pegasys/teku/storage/server/rocksdb/V3RocksDbDatabaseTest.java (hunk 2 above)
Comment (cemozerr) [reply storage/src/test/java/tech/pegasys/teku/storage/server/rocksdb/V3RocksDbDatabaseTest.java]: done.

File: storage/src/main/java/tech/pegasys/teku/storage/server/rocksdb/dataaccess/V3RocksDbDao.java (hunk 3 above)
Comment (cemozerr) [reply storage/src/main/java/tech/pegasys/teku/storage/server/rocksdb/dataaccess/V3RocksDbDao.java]: makes sense.

File: storage/src/test/java/tech/pegasys/teku/storage/server/rocksdb/V3RocksDbDatabaseTest.java (hunk 4 above)
Comment (cemozerr) [reply storage/src/test/java/tech/pegasys/teku/storage/server/rocksdb/V3RocksDbDatabaseTest.java]: Thanks good spot.

File: storage/src/test/java/tech/pegasys/teku/storage/server/rocksdb/V3RocksDbDatabaseTest.java (hunk 5 above)
Comment (cemozerr) [reply storage/src/test/java/tech/pegasys/teku/storage/server/rocksdb/V3RocksDbDatabaseTest.java]: removed.

File: storage/src/test/java/tech/pegasys/teku/storage/server/rocksdb/V3RocksDbDatabaseTest.java (hunk 6 above)
Comment (cemozerr) [reply storage/src/test/java/tech/pegasys/teku/storage/server/rocksdb/V3RocksDbDatabaseTest.java]: removed.

**Pull request merged**